Document how to build and use deployment artifacts (#5215)
cfpb/cfgov-refresh: docs/deployment.md

> This repository includes code for generating a self-zip archive

Reviewer: What does "self-zip" mean?

Author: Andy caught that-- I meant to write self-extracting

> This repository includes code for generating a self-zip archive

Reviewer 2: ```suggestion
This repository includes code for generating a self-extracting zip archive
```

> of the code and all of it's Python dependencies. We use these

Reviewer: ```suggestion
of the code and all of its Python dependencies. We use these
```

> of the code and all of it's Python dependencies. We use these

Reviewer 2: ```suggestion
of the code and all of its Python dependencies. We use these
```

archives to deploy the site to a Linux server.

> # What's an artifact?

Reviewer: ```suggestion
## What's an artifact?
```

> # What's an artifact?

Reviewer 2: I propose moving this section down below the "Generating" section, as, at first read, this provides an awful lot of detail without first getting a sense for the purpose of the artifact at a high level.

Author: This is done in the latest commit

> It's a zip file, that is executable and includes a Python "shebang" line. Here's a

Reviewer: What's the importance of the fact that it includes a Python "shebang" line? This might be self-explanatory to most reading this doc, but it might also be helpful to elaborate on that.

Author: It's the thing that makes it self extracting-- if it didn't have a shebang line, it wouldn't be a valid executable.

Author: I re-arranged that section a bit, to describe that a little better.

(very abbreviated) peek into what's *in* the zip file:

```
__main__.py
cfgov.pth
loadenv.py
setup.py
loadenv-init.pth
wheels/botocore-1.10.84-py2.py3-none-any.whl
wheels/tqdm-4.15.0-py2.py3-none-any.whl
...
static.in/0/
static.in/0/fonts/
static.in/0/fonts/8344e877-560d-44d4-82eb-9822766676f9.woff
...
bootstrap_wheels/setuptools-41.2.0-py2.py3-none-any.whl
bootstrap_wheels/pip-19.2.3-py2.py3-none-any.whl
cfgov/...
```

> The `__main__.py` file contains [the code that runs when the zipfile is invoked

Reviewer: ```suggestion
The `__main__.py` file contains [the code that runs when the zip file is invoked
```

> as a python module](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/deployable_zipfile/extract.py)

Reviewer: ```suggestion
as a Python module](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/deployable_zipfile/extract.py)
```

> as a python module](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/deployable_zipfile/extract.py)

Reviewer 2: ```suggestion
as a Python module](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/deployable_zipfile/extract.py)
```

> The `wheels/` directory contains all of our python dependencies, while

Reviewer: ```suggestion
The `wheels/` directory contains all of our Python dependencies, while
```

> `bootstrap_wheels` contains modules needed at deployment time, to install

Reviewer 2: ```suggestion
`bootstrap_wheels/` contains modules needed at deployment time, to install
```

everything.

`loadenv.py` and `loadenv-init.pth` are used to load environment variables from

> a `environment.json` file, deployed seperately.

Reviewer: ```suggestion
an `environment.json` file, deployed seperately.
```

> # Generating a deployment artifact

Reviewer: ```suggestion
## Generating a deployment artifact
```

Running the script at `./docker/deployable-zipfile/build.sh` will start a CentOS 6

> container, generate the artifact, and save it to `./cfgov_current_build.zip`.

Reviewer 2: Could you add a link to [the script](https://github.com/cfpb/cfgov-refresh/blob/master/docker/deployable-zipfile/_build.sh) that generates the artifact?

Also, it'd be nice to add an explanation here about why the artifact needs to be generated on CentOS.

Author: I've added the link

> # Deploying an artifact

Reviewer: ```suggestion
## Deploying an artifact
```

We currently use Ansible to prepare servers to run an artifact, and to do the actual deployment.

> If we ignore some specifics and quirks of our environment, the basic steps look something like this:

Reviewer 2: Maybe add a comment about how these scripts are maintained in a private repository?

> If we ignore some specifics and quirks of our environment, the basic steps look something like this:

Reviewer: Minor suggestion: It might be nice to make the list below a numbered list.

Author: I've made this change

- copy the artifact to the system

> - execute the artifact, with the -d (destination) argument, `./artifact.zip -d destination-dir`. This

Reviewer: ```suggestion
- Execute the artifact, with the `-d (destination)` argument: `./artifact.zip -d destination-dir`. This
```

Please also add spaces to the start of each line below to visually indicate in the code that they're part of this bullet.

> will unpack the files, create a new virtualenv, install all of the wheels in `wheels/` into that

Reviewer 2: ```suggestion
will unpack the files, create a new virtualenv, and install all of the wheels in `wheels/` into that
```

> virtualenv. **Important Note:** This should be done with the same python interpreter that will run the

Reviewer 2: ```suggestion
virtualenv. **Important Note:** This should be done with the same Python interpreter that will run the
```

> virtualenv. **Important Note:** This should be done with the same python interpreter that will run the

Reviewer: ```suggestion
virtualenv. **Important Note:** This should be done with the same Python interpreter that will run the
```

application. For example, on our servers this means using [`scl enable`](https://linux.die.net/man/1/scl)
to specify a particular Python version from 
[Software Collections](https://www.softwarecollections.org/en/scls/?search=python).

> - put an 'environment.json' file in place, in your destination-dir

Reviewer: ```suggestion
- put an `environment.json` file in place, in your `destination-dir`
```

> - run django utilities like 'collectstatic' and 'migrate'

Reviewer 2: ```suggestion
- run Django utilities like 'collectstatic' and 'migrate'
```

> - run django utilities like 'collectstatic' and 'migrate'

Reviewer: ```suggestion
- run Django utilities like `collectstatic` and `migrate`
```

This repository includes code for generating a self-extracting zip archive
of the code and all of its Python dependencies. We use these
archives to deploy the site to a Linux server.

# Generating a deployment artifact

Running the script at `./docker/deployable-zipfile/build.sh` will start a CentOS 6
container, generate the artifact, and save it to `./cfgov_current_build.zip`.

We use CentOS 6 here, so that the Python modules that include compiled code, will
be compiled for the same environment they will be run in.

# What's in an artifact?

It's a zip file, that is executable and includes a Python "shebang" line. Here's a
(very abbreviated) peek into what's *in* the zip file:

```
__main__.py
cfgov.pth
loadenv.py
setup.py
loadenv-init.pth
wheels/botocore-1.10.84-py2.py3-none-any.whl
wheels/tqdm-4.15.0-py2.py3-none-any.whl
...
static.in/0/
static.in/0/fonts/
static.in/0/fonts/8344e877-560d-44d4-82eb-9822766676f9.woff
...
bootstrap_wheels/setuptools-41.2.0-py2.py3-none-any.whl
bootstrap_wheels/pip-19.2.3-py2.py3-none-any.whl
cfgov/...
```

The `__main__.py` file contains [the code that runs when the zip file is invoked

> as a Python module](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/deployable_zipfile/extract.py)

Reviewer 2: ```suggestion
as a Python module](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/deployable_zipfile/extract.py).
```

The `wheels/` directory contains all of our python dependencies, while
`bootstrap_wheels/` contains modules needed at deployment time, to install
everything.

`loadenv.py` and `loadenv-init.pth` are used to load environment variables from
a `environment.json` file, deployed seperately.

# Deploying an artifact

We currently use Ansible to prepare servers to run an artifact, and to do the actual deployment.
If we ignore some specifics and quirks of our environment, the basic steps look something like this:

1. copy the artifact to the system

> 2. Execute the artifact, with the -d (destination) argument, `./artifact.zip -d destination-dir`. This

Reviewer 2: ```suggestion
2. execute the artifact, with the -d (destination) argument, `./artifact.zip -d destination-dir`. This
```

(just to be consistent with capitalization)
